src/DueDateHistogram.ts

```diff
@@ -0,0 +1,137 @@
+import { Card } from "./Card";
+import { Deck } from "./Deck";
+import {
+    CardOrder,
+    DeckOrder,
+    DeckTreeIterator,
+    IDeckTreeIterator,
+    IIteratorOrder,
+} from "./DeckTreeIterator";
+import { NoteReviewDeck, SchedNote } from "./NoteReviewDeck";
+import { TopicPath } from "./TopicPath";
+import { RepItemScheduleInfo } from "./algorithms/base/RepItemScheduleInfo";
+import { OsrNoteGraph } from "./algorithms/osr/OsrNoteGraph";
+import { TICKS_PER_DAY } from "./constants";
+import { globalDateProvider } from "./util/DateProvider";
+
+export class DueDateHistogram {
+    // The key for dueDatesNotes is the number of days after today
+    // therefore the key to lookup how many cards are due today is 0
+    public static dueNowNDays: number = 0;
+
+    // Key - # of days in future
+    // Value - Count of notes due
+    dueDatesMap: Map<number, number> = new Map<number, number>();
+
+    constructor(rec: Record<number, number> = null) {
+        this.dueDatesMap = new Map<number, number>();
+        if (rec != null) {
+            Object.entries(rec).forEach(([key, value]) => {
+                this.dueDatesMap.set(Number(key), value);
+            });
+        }
+    }
+
+    get dueNotesCount(): number {
+        let result: number = 0;
+        if (this.dueDatesMap.has(DueDateHistogram.dueNowNDays))
+            result = this.dueDatesMap.get(DueDateHistogram.dueNowNDays);
+        return result;
+    }
+
+    hasEntryForDays(days: number): boolean {
+        return this.dueDatesMap.has(days);
+    }
+
+    set(days: number, value: number): void {
+        this.dueDatesMap.set(days, value);
+    }
+
+    get(days: number): number {
+        return this.dueDatesMap.get(days);
+    }
+
+    increment(days: number): void {
+        let value: number = 0;
+        if (this.dueDatesMap.has(days)) {
+            value = this.dueDatesMap.get(days);
+        }
+        this.dueDatesMap.set(days, value + 1);
+    }
+
+    decrement(days: number): void {
+        let value: number = 0;
+        if (this.dueDatesMap.has(days)) {
+            value = this.dueDatesMap.get(days);
+            this.dueDatesMap.set(days, value - 1);
+        }
+    }
```

Comment:
More like a question, but why the "asymmetry" between `increment` & `decrement`... `increment` does the `.set` outside the `if` but `decrement` does it inside

Reply:
It is valid for `increment` to be called when there is no entry in the map related to `days`, in which case we need to set the value to 1 outside the `if`.

But it's not valid for `decrement` to be called in that situation, because we should never be setting the value to -1. I didn't want to throw an exception and just guarded it with the `if`. I'll modify the code to also catch the case of `decrement` being called where `dueDatesMap.get(days)` has a value of 0, and not just if `dueDatesMap.has(days)` is `false`.